src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        void IList.Insert(int index, object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.Insert(int index, T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.Remove(object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        bool ICollection<T>.Remove(T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+            => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => GetEnumerator();
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other is null)
+                return 1;
+
+            if (!(other is SegmentedArray<T> o)
+                || Length != o.Length)
+            {
+                throw new ArgumentException(CompilerExtensionsResources.ArgumentException_OtherNotArrayOfCorrectLength, nameof(other));
+            }
+
+            for (var i = 0; i < Length; i++)
+            {
+                var result = comparer.Compare(this[i], o[i]);
+                if (result != 0)
+                    return result;
+            }
+
+            return 0;
+        }
+
+        bool IStructuralEquatable.Equals(object other, IEqualityComparer comparer)
+        {
+            if (other is null)
+                return false;
+
+            if (!(other is SegmentedArray<T> o))
+                return false;
+
+            if ((object)_items == o._items)
+                return true;
+
+            if (Length != o.Length)
+                return false;
+
+            for (var i = 0; i < Length; i++)
+            {
+                if (!comparer.Equals(this[i], o[i]))
+                    return false;
+            }
+
+            return true;
+        }
+
+        int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
+        {
+            _ = comparer ?? throw new ArgumentNullException(nameof(comparer));
+
+            var ret = 0;
+            for (var i = Length >= 8 ? Length - 8 : 0; i < Length; i++)
```

Comment:
doc.

Reply:
slightly wary of this.  note that java moved away from thsi approach because they did run into common cases where hashes tried to eb smart, but people would common run into that were problematic.  i.e. for us, it would be if people had different elements at the start, but the end of the array was all default/null.  this could get lots of collisions.

Reply:
https://github.com/dotnet/runtime/blob/e0ec035994179e8ebd6ccf081711ee11d4c5491b/src/libraries/System.Private.CoreLib/src/System/Array.cs#L380-L383

Reply:
doc please